homeassistant/components/toon/__init__.py

```diff
@@ -40,49 +73,119 @@
+        identifiers={(DOMAIN, toon.agreement.id, 'meter_adapter')},
         manufacturer='Eneco',
         name="Meter Adapter",
-        via_hub=(DOMAIN, toon.agreement.id)
+        via_hub=(DOMAIN, toon.agreement.id),
+    )
+
+    def update(call):
+        """Service call to manually update the data."""
+        called_display = call.data.get(CONF_DISPLAY, None)
+        for toon_data in hass.data[DATA_TOON].values():
+            if (called_display and called_display == toon_data.display_name) \
+                    or not called_display:
+                toon_data.update()
+
+    hass.services.async_register(
+        DOMAIN, "update", update, schema=SERVICE_SCHEMA
     )
 
     for component in 'binary_sensor', 'climate', 'sensor':
         hass.async_create_task(
-            hass.config_entries.async_forward_entry_setup(entry, component))
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
 
     return True
 
 
+class ToonData:
+    """Communication class for interacting with toonapilib."""
+
+    def __init__(
+            self,
+            hass: HomeAssistantType,
+            entry: ConfigType,
+            toon
+    ):
+        """Initialize the Toon data object."""
+        self._hass = hass
+        self._toon = toon
+        self._entry = entry
+        self.agreement = toon.agreement
+        self.gas = toon.gas
+        self.power = toon.power
+        self.solar = toon.solar
+        self.temperature = toon.temperature
+        self.thermostat = toon.thermostat
+        self.thermostat_info = toon.thermostat_info
+        self.thermostat_state = toon.thermostat_state
+
+    @property
+    def display_name(self):
+        """Return the display connected to."""
+        return self._entry.data[CONF_DISPLAY]
+
+    def update(self, now=None):
+        """Update all Toon data and notify entities."""
+        # Ignore the TTL meganism from client library
+        # It causes a lots of issues, hence we take control over caching
+        self._toon._clear_cache()  # noqa  pylint: disable=W0212
+
+        # Gather data from client library (single API call)
+        self.gas = self._toon.gas
+        self.power = self._toon.power
+        self.solar = self._toon.solar
+        self.temperature = self._toon.temperature
+        self.thermostat = self._toon.thermostat
+        self.thermostat_info = self._toon.thermostat_info
+        self.thermostat_state = self._toon.thermostat_state
+
+        # Notify all entities
+        dispatcher_send(
+            self._hass, DATA_TOON_UPDATED, self._entry.data[CONF_DISPLAY]
+        )
+
+
 class ToonEntity(Entity):
     """Defines a base Toon entity."""
 
-    def __init__(self, toon, name: str, icon: str) -> None:
+    def __init__(self, toon: ToonData, name: str, icon: str) -> None:
         """Initialize the Toon entity."""
         self._name = name
         self._state = None
         self._icon = icon
         self.toon = toon
+        self.update()
```

Comment:
use `add_entities([], True)`

Reply:
🤦‍♂ Not sure what I was thinking... Addressed, thanks!